proposed/logging-generator.md

```diff
@@ -0,0 +1,327 @@
+        return true;
+    }
+
+    public KeyValuePair<string, object?> this[int index]
+    {
+        get => ... // depends on the type used to store key value pairs
+    }
+
+    public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+    {
+        for (int i = 0; i < Count; ++i)
+        {
+            yield return this[i];
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+}
+```
+
+In order to make use of a builder like `CustomLoggerParamsBuilder`, we would need new overloads in the future for LogInformation (and other similar APIs):
+
+```c#
+// We need new overloads to support this call:
+logger.LogInformation($"Welcome to {city}!");
+```
+
+But the sample below shows an application of the builder using an existing logging API:
+```c#
+var city = "Vancouver";
+
+logger.Log(
+    LogLevel.Information,
+    new EventId(1),
+    $"Welcome to {city}!",
+    new Exception(),
+    (s, e) => s);
+```
+The compiler would translate the above call to:
+```c#
+logger.Log<CustomLoggerParamsBuilder>(
+    LogLevel.Information,
+    new EventId(1),
+    builder, // the CustomLoggerParamsBuilder
+    new Exception(),
+    builder.FormatCallback
+);
+```
+Through this new compiler feature, the builder would be able to recognize name holes, (here the city) or even format specifiers. However, currently the proposed builder approach is limited in its ability to get both name holes and format specifiers at once. The logger to is able to identify both of these from message templates and would be good to have this feature available.
+
+Also, an older compiler, the builder overload may not be detectable and therefore the above log APIs could end up calling the wrong overload, which would cause in logging messages to completely lose the structure provided in the input. This could be however be mitigated with an analyzer that warns user when they use string interpolated messages on older compiler versions.
+ 
+### Limitations with string interpolation builder approach:
+
+- With the interpolated string builder approach, there is currently no way to get both name holes and format specifiers, but this is supported today already by the existing logger APIs.
+
+- The biggest concern for using the custom struct builder for our logging approach is that we'd need to store some value for name holes eagerly in the builder. This step cannot be done lazily because the structure we generate could ultimately get serialized in different ways decided by the different consumers of `ILogger.Log` call. The end result of using the builder this way, could be less efficient than what we already have in our logging APIs, even though our current approaches already do boxing.
+
+### Benefits:
+
+- The name hole approach adds enough value that would be good to consider in the future for cases where we are given an interpolated string rather than message templates.
+
```

Comment:
Also, the comment made back on the original issue stated:
"A benefit there is that we can use string format syntax as a natural way which we can consider for imperative logging, but the consensus there is to keep the logging generator to enable writing more declarative logging APIs."
FWIW, I think this is a strange line being drawn between "declarative" and "imperative".  From the developer perspective, we're talking about the difference between (using one of your examples from earlier):
```C#
[LoggerMessage(0, LogLevel.Information, "Hello `{name}`")]
public static partial void LogName(this ILogger logger, string name);
```
and
```C#
public static void LogName(this ILogger logger, string name) =>
    logger.Log(0, LogLevel.Information, $"Hello `{name}`");
```
If we define "declarative" as using an attribute, sure, one does and one doesn't.  But otherwise, we're still declaring the same information, just slightly different syntax.

The valid reasons I heard that we might prefer the source generator approach were:
- Wanting to enable the exact format syntax required by https://messagetemplates.org/
- Wanting to be able to pass all of original data as-is without any complication around how it's stored prior to something being done with it other than creating a string
- Wanting to be able to supply alternative names for the holes, if the C# 10 syntax doesn't end up allowing that (but Fred is investigating)

None of the other reasons I heard were compelling to me for why we'd go the source generator route.

Reply:
That's fair. I think they both offer declarative. I will update the proposal.

Reply:
I honestly don't think these approaches are at odds but:  
- The structure *has* to be preserved (Names), that's the main thing about the string builder approach that is missing (hopefully we can make something work in C# 10).
- Not being able to lazily ToString is a problem, we don't want to allocate until a logger provider asks for the string
- The approach is declarative and toolable. We can write rules against the attribute instead of call sites that use Log.

I say we treat the builder approach as distinct from the source generator. In fact, in the later proposals we're going to generate calls to `LoggerMessage.Define<...>`. This change basically re-enforces that the source generator is a declarative approach to the model we have today.

Reply:
> The structure has to be preserved (Names), that's the main thing about the string builder approach that is missing

This is my "Wanting to be able to supply alternative names for the holes" bullet.  If just `{name}` is used, it can already support that.  I suggest if specifying names separate from the expression itself is important, that the design document call that out and show it in examples: currently I don't see any examples that supply different names.

> Not being able to lazily ToString is a problem, we don't want to allocate until a logger provider asks for the string

This is my "Wanting to be able to pass all of original data as-is without any complication around how it's stored prior to something being done with it other than creating a string" bullet. But it depends on what you mean by allocating.  I imagine the builder would either format into an ArrayPool-rented char[] or serialize into an ArrayPool-rented byte[].  And of course only after checking whether logging was enabled at an appropriate logging level.

> The approach is declarative and toolable

I still question the declarative statement here.  An analyzer would search for attributes and look at attribute arguments, or an analyzer would search for Log method calls and look at method call arguments.  Is the statement really that sites are implicitly constrained by what you're allowed to write in attributes?

Reply:
From my perspective "declarative" is really about the fact it is a "guided" experience. The code generator has deep semantic understanding of what the user is doing and provides useful error messages that don't exist using completely generic mechanisms.

For example, the code generator generates warnings for duplicate event ids. It'll auto-generate a message if you don't supply one to output parameters as a JSON blob, etc.

Reply:
In terms of supplying different names, the current implementation provides a limited form of that as it optionally supports pascal casing of names. So you can supply "foo" as a method parameter, and call it "{Foo}" in the message text (and in the downstream IReadOnlyCollection produced on the TState given to ILogger). This feature is controlled by a global option on the generator.

Reply:
> This is my "Wanting to be able to supply alternative names for the holes" bullet. If just {name} is used, it can already support that. I suggest if specifying names separate from the expression itself is important, that the design document call that out and show it in examples: currently I don't see any examples that supply different names.

We change casing all of the time in ASP.NET Core. We also define templates separately from where we specify values so this happens *all the time*. The intent isn't to capture variables but to design log messages and their structure and pass the appropriate values after the fact.

> This is my "Wanting to be able to pass all of original data as-is without any complication around how it's stored prior to something being done with it other than creating a string" bullet. But it depends on what you mean by allocating. I imagine the builder would either format into an ArrayPool-rented char[] or serialize into an ArrayPool-rented byte[]. And of course only after checking whether logging was enabled at an appropriate logging level.

We need to eagerly stash all of the state before we ToString the overall thing. This makes it hard to avoid boxing or calling ToString on the incoming parameters. So there's an allocation forced up front to produce the builder.

> I still question the declarative statement here. An analyzer would search for attributes and look at attribute arguments, or an analyzer would search for Log method calls and look at method call arguments. Is the statement really that sites are implicitly constrained by what you're allowed to write in attributes?

Attributes force a particular pattern that we like and is easier to analyze than code. This is a style preference thing though.

Reply:
> We need to eagerly stash all of the state before we ToString the overall thing. This makes it hard to avoid boxing or calling ToString on the incoming parameters. So there's an allocation forced up front to produce the builder.

If you don't count ArrayPool rentals as allocation (waving my hands), then there's no allocation to create the builder.  You're right there would be ToString allocations as part of formatting into that, for any non-primitive types that allocate in their ToString methods. I expect with ISpanFormattable that will go way down, though.